_posts/2021-11-18-shrinking-size-of-monorepo.md

```diff
@@ -0,0 +1,133 @@
+`.dylib`, `.pdf`, `.zip`, `.mp4`, `.json`
+
+We decided to track BLOBs using [Git LFS \(Large File Storage\)](https://git-lfs.github.com/). In our case the
+following were a good use case for it:
+- large binary files `.jpg`, `.png`, `.a`, `.dylib`, `.pdf`, `.zip`, `.mp4`, `.json`
+- framework binaries
+- toolset binaries
+
+### Reproducible procedure and dry runs
+
+We created a script that contained all commands that removed redundant items from the history. To remove items we used
+[git filter-repo](https://github.com/newren/git-filter-repo) - it‘s much more performant than git‘s built-in
+ `git filter-branch` (do not use it!). Some examples of usage:
+
+```shell
+git filter-repo --invert-paths --path Pods/ --force
+git filter-repo --invert-paths --paths-from-file remove.txt --force
+git filter-repo --invert-paths --path-glob '*.pbxproj' --force
+
+```
+
+After the removal we restored the most-recent version of binaries, frameworks and BLOBs to the repo and we tracked them
+with Git LFS:
+
+```shell
+git lfs track "*.png"
+git lfs track "Vendor/SomeSDK/SomeSDK.framework/SomeSDK"
+
+```
+
+We ran the script a few times to verify the output size of the repo. One crucial aspect after the run was to verify
+that **all plans** on the CI (Continuous Integration) pass - we did it to check that the app still compiles, tests pass
+and no more files that the ones we had wanted were actually deleted.
+
+### Communication
+
+The crucial aspect of introducing any change is communication. It‘s good to prepare it in advance, have team members
+review it. We used a few channels so that our devs would get important info about the migration and history rewrite
+through the channel that suited their working habits best (e-mails, instant messaging tool, dev forums).
+
+## Some final thoughts
+
+It took a large amount of time to prepare the migration, understand the history of the repository and select proper
+items and
+strategies for the migration. The links here might be a good starting point if you wanted to rewrite histories of your
+overweight repos:
+
+- [GitHub Universe Workshops](https://www.youtube.com/watch?v=bk7akV8nyAM) and corresponding
+[repository](https://github.com/githubuniverseworkshops/grafting-monorepos)
+- [git-sizer](https://github.com/github/git-sizer)
+- [git filter-repo](https://github.com/newren/git-filter-repo)
+
+When creating a plan for the rewrite, remember to have a checklist that you can use to verify outcomes and to remember
+all the steps involved in the process. If the repository migration from one provider to another hosting provider is
+required execute it together with the history rewrite. Plan the rewrite for a time that folks would not want to
+push code to the repo. We used Friday evening, and yes, we had to fix some issues over the weekend - not everything
+went smoothly.
+
+You can have a copy of your old repository in READ-ONLY mode on the servers - it will serve as a **backup** and will
```

Comment:
```suggestion
Finally, it is useful to have a copy of your old repository in READ-ONLY mode on the servers – it will serve as a **backup** and will
```